hazelcast/src/main/java/com/hazelcast/internal/networking/ChannelHandler.java

```diff
@@ -0,0 +1,140 @@
+ * protocol to write. But this should not be dealt with by the TldHandshakeEncoder,
+ * because it will mess up the communication with the other side.
+ *
+ * Perhaps currently not a problem because in case of the ProtocolEncoder, it
+ * will copy the data to the PacketHandler, not to the TlsHandshakeEncoder.
+ *
+ * Replacing the source/destination of a Handler after the Handler is registered
+ * is not adviced, it can lead to src/destination changes not being picked up
+ * because the linking process will not be repeated.
+ *
+ *
+ *
+ *
+ * @param <S> the type of the source. E.g. a ByteBuffer or
+ *           a {@link com.hazelcast.util.function.Supplier}.
+ * @param <D> the type of the destination. E.g. a ByteBuffer
+ *           or a {@link com.hazelcast.util.function.Consumer}.
+ */
+public abstract class ChannelHandler<H extends ChannelHandler, S, D> {
+
+    /**
+     * The Channel this InboundHandler is connected to. Only the {
+     * @link ChannelInboundPipeline} should modify this field.
+     */
+    protected Channel channel;
+
+    protected S src;
+    protected D dst;
+
+    /**
+     * Gets the source of this ChannelHandler.
+     *
+     * This method should only be called from the thread that owns this ChannelHandler.
+     *
+     * @return the source; could be null.
+     */
+    public S src() {
+        return src;
+    }
+
+    /**
+     * Sets the source of this ChannelHandler.
+     *
+     * This method should only be called from the thread that owns this ChannelHandler.
+     *
+     * @param src the new source; is allowed to be null.
+     */
+    public void src(S src) {
+        this.src = src;
+    }
+
+    /**
+     * Gets the destination of this ChannelHandler.
+     *
+     * This method should only be called from the thread that owns this ChannelHandler.
+     *
+     * @return the destination; could be null.
+     */
+    public D dst() {
+        return dst;
+    }
+
+    /**
+     * Sets the destination of this ChannelHandler.
+     *
+     * This method should only be called from the thread that owns this ChannelHandler.
+     *
+     * @param dst the new destination; is allowed to be null.
+     */
+    public void dst(D dst) {
+        this.dst = dst;
+    }
+
+    /**
+     * Sets the Channel.
+     *
+     * Should only be called by the {@link ChannelInboundPipeline}.
+     *
+     * @param channel the Channel
+     * @return this instance
+     */
+    public final H setChannel(Channel channel) {
+        this.channel = channel;
+        return (H) this;
+    }
+
+    /**
+     * Gets called when this ChannelHandler is added to the {@link ChannelInboundPipeline}.
+     *
+     * This method should execute very quickly because it could be executed on an io thread.
+     *
+     * Will be called from a thread owning this handler.
+     */
+    public void handlerAdded() {
+    }
+
+    public void channelInactive() throws Exception {
```

Comment:
What `channelInactive()` is for? Can you add some doc?

Reply:
One part of the design I'm not 100% sure of. It is related to TLS goodbye handshake.